pkg/image/origins/main.go

```diff
@@ -0,0 +1,284 @@
+	if err != nil {
+		return err
+	}
+	// look through the windows target images
+	err = convertImagesToRepoUrls(targetWindowsImages, imgToURL)
+	if err != nil {
+		return err
+	}
+
+	// order images alphabetically
+	type sorter struct {
+		img string
+		url string
+	}
+	var sorted []sorter
+	for k, v := range imgToURL {
+		sorted = append(sorted, sorter{img: k, url: v})
+	}
+
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].img < sorted[j].img
+	})
+
+	x := "OriginMap = map[string]string{\n"
+	for _, e := range sorted {
+		x = x + "	\"" + e.img + "\":" + " " + "\"" + e.url + "\"," + "\n"
+	}
+	x = x + "}\n"
+
+	// print result, can be copied into pkg/image/origins.go
+	// to automatically update some images.
+	fmt.Println(x)
+
+	// find all images that have not been resolved yet
+	var unknownImages string
+	for k, v := range imgToURL {
+		if v == "" || v == "unknown" {
+			unknownImages = unknownImages + "\n" + k
+		}
+	}
+
+	// warn about unresolved images
+	// so that they may be manually resolved
+	if len(unknownImages) > 0 {
+		fmt.Println(fmt.Sprintf("[WARN] Some images do not have an origin defined, please provide origins within rancher/pkg/image/origins.go for the following images: %s", unknownImages))
+	}
+
+	return nil
+}
+
+// convertImagesToRepoUrls attempts to resolve Rancher owned images with its origin repository.
+// It does so on a best-effort basis, any gaps left by this must be covered manually
+// in the OriginMap in origins.go
+func convertImagesToRepoUrls(images []string, imgToURL map[string]string) error {
+	for _, repo := range img.UniqueTargetImages(images) {
+		if v, ok := imgToURL[repo]; ok && (v != "" && v != "unknown") {
+			continue // image is already resolved
+		}
+		switch {
+		case strings.Contains(repo, "mirrored"):
+			// no way to automatically determine origin.
+			// needs manual resolution, see origins.go.
+			imgToURL[repo] = "unknown"
+		case strings.Contains(repo, "hardened"):
+			// hardened images conversion, best effort
+			cleanRepo := strings.ReplaceAll(repo, "hardened-", "image-build-")
+			url, err := checkURL("https://www.github.com/rancher/" + cleanRepo)
+			if err != nil && err.Error() != imageNotFound {
+				return err
+			}
+			if err != nil {
+				url = "unknown"
+			}
+			imgToURL[repo] = url
+		default:
+			// handle Rancher images which share
+			// a name with their base repository
+			url, err := checkURL("https://www.github.com/rancher/" + repo)
+			if err != nil && err.Error() != imageNotFound {
+				return err
+			}
+			if err != nil {
+				url = "unknown"
+			}
+			imgToURL[repo] = url
+		}
+	}
+	return nil
+}
+
+// checkURL performs a GET request against a github repository
+// to determine if it exists or not.
+func checkURL(url string) (string, error) {
+	r, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	if r.StatusCode != http.StatusOK {
+		return "", fmt.Errorf(imageNotFound)
+	}
+	return url, nil
+}
+
+func GetWindowsAgentImage() string {
```

Comment:
The same function is already defined in `pkg/image/export/main.go`. 
How about creating a new file `pkg/image/utilities.go` for all common helper functions?

Reply:
Created `utilities.go` to handle shared logic 👍